docs/proposals/event-notification.md

```diff
@@ -0,0 +1,141 @@
+
+In short, the proposal is:
+- New flag for push command, `odo push -o json`: Outputs JSON events that correspond to what devfile actions/commands are being executed by push.
+- New odo command `odo component status -o json`: Pings the application URLs, and checks the container/pod status every X seconds (and/or a watch, for Kubernetes case). The result of both is output as JSON to be used by external tools to determine if the application is running.
+- A standardized markup format for communicating detailed build status, something like: `#devfile-status# {"buildStatus":"Compiling application"}`, to be optionally included in devfile scripts that wish to provide detailed build status to consuming tools.
+
+## Terminology
+
+With this issue, we are looking at adding odo support for allowing external tools to gather build status and application status. We further divide both statuses into detailed and non-detailed, with detailed principally being related to statuses that can be determine by looking at container logs.
+
+**Build status**: A simple build status indicating whether the build is running (y/n), and whether the last build succeeded or failed. This can be determined based on whether odo is running a dev file action/command, and the process error code (0 = success, non-0 = failed) of that action/command.
+
+**Detailed build status**: An indication of which step in the build process the build is in. For example: are we 'Compiling application', or are we 'Running unit tests'?
+
+**App status**: Determine if an application is running, using container status (for both local and Kubernetes, various status: containercreating, started, restarting, etc), or whether an HTTP/S request to the root application URL returns an HTTP response with any status code.
+
+**Detailed application status**: 
+- While app status works well for standalone application frameworks (Go, Node Express, Spring Boot), it works less well for full server runtimes such as Java EE application servers like OpenLiberty/WildFly that may begin responding to Web requests before the user's deployed WAR/EAR application has finished startup. 
+- Since these application servers are built to serve multiple concurrently-deployed applications, it is more difficult to determine the status of any specific application running on them. The lifecycle of the application server differs from the lifecycle of the application running inside the application server. 
+- Fortunately, in these cases we can use the console logs (from `odo log`) from the runtime container to determine a more detailed application status.
+- For example, OpenLiberty (as an example of an application-server-style container) prints a specific code when an application is starting  `CWWKZ0018I: Starting application {0}.`, and another when it has started. `CWWKZ0001I: Application {0} started in {1} seconds.`
```

Comment:
How would odo know what log line means what? For example that `CWWKZ0018I: Starting application {0}.` means that the application is starting initialization. This is framework-specific knowledge and as such, it should be somewhere in devfile or in a Stack but not inside odo.

Reply:
Yep, we definitely don't want ODO to know anything about these codes/strings, it looks like the language I used in the proposal is imprecise re: who would use these data. 

I have updated the proposal to clarify that it is the consuming tool's (for example, IDE) responsibility to know what these codes mean.. I added these lines:
> Odo itself should NOT know anything about these specific application codes; knowing how these translate into a detailed application status would be the responsibility of the IDE/consuming tool. Odo's role here is only to provide the console output log. 
>
> In the future, we could add these codes into the devfile to give Odo itself some agency over determining the detailed application status, but for this initial proposal I am leaving this responsibility with the consuming tool.
```

Reply:
> Yah, we definitely don't want odo to know anything about these codes/strings; it looks like the language I used in the proposal is imprecise re: who would use these data.
> 
> I have updated the proposal to clarify that it is the _consuming tool_'s responsibility (for example, IDE) to know what these codes mean.. I added these lines:

Maybe it was just my bad implementation of this. What you added makes this perfectly clear, thx.